### DISCUSSION TITLE
Optional command args

### DISCUSSION THREAD

**Discussion by jbakosi** [daniele77/cli#150] [Q&A]
Is it possible to have optional command arguments? What I'd like to do is the following.

Instead:
```
 - connect <string> <int>
        Connect to sever, given by hostname and port. 
```
I would like
```
 - connect [hostname[:port]]
        Connect to sever, given by optional hostname and optional port. 
```
So my question is two-fold:
1. Is there a way to have optional arguments to commands?
2. Can the above help string be displayed instead of the more generic `<string> <int>`?

Thanks.

**Comment by daniele77** [comment]
Hi @jbakosi ,
I need to answer your questions in reverse order :-)

**Answer to question 2**:  you can specify the description of the parameters by using the syntax:

        rootMenu->Insert(
                "add", {"first_term", "second_term"},  // <==== this optional parameter is the argument description
                [](std::ostream& out, int x, int y)
                {
                    out << x << " + " << y << " = " << (x+y) << "\n";
                },
                "Print the sum of the two numbers" );


**Answer to question 1**:  the library doesn't allow real "optional parameters", however, you can repeat the same command several times (see the command "add" in the "complete.cpp" example):

        rootMenu->Insert(
                "add", {"first_term", "second_term"},
                [](std::ostream& out, int x, int y)
                {
                    out << x << " + " << y << " = " << (x+y) << "\n";
                },
                "Print the sum of the two numbers" );
        rootMenu->Insert(
                "add", {"first_term", "second_term", "third_term"},
                [](std::ostream& out, int x, int y, int z)
                {
                    out << x << " + " << y << " + " << z << " = " << (x+y+z) << "\n";
                },
                "Print the sum of the three numbers" );

in this case, the "z" parameter is sort of optional. You can use this feature to mimic optional parameters, however, this results in having the same command repeated in the help description.

**Comment by jbakosi** [comment]
Thanks for the answers @daniele77. 3 is only a workaround and it may be confusing to list the command twice in the help, and 2 encloses the string passed in angled brackets which are usually understood as mandatory, so it may also be confusing.

How much effort do you think would it take to add support for truly optional arguments? If not too bad, I could look into it and try to add it (with potentially some help from you.)

**Reply by daniele77** [reply]
First of all, sorry for the bad numeration (I wanted to do 2. and 1. but github autocompleted them in progressive order).

Anyway.

You're right with both the answers. Actually, cli adds angle brackets to parameter description (because, you know, there is no support for optional parameters in the library :-) ) . 

Maybe a third option could be: using a ` vector<string>` as a parameter and then making all the checks and conversions inside your callback. You can check if the size of the vector is 2 or 1, and in that case, you know whether the user provided the optional 2nd parameter or not. However, I guess you still get the angle brackets in the parameter description.

You can try to add optional arguments in the library, but I'm afraid it's quite difficult (if not impossible), due to the parsing engine of the commands. Anyway, if you want to give it a try, the code you should modify is in cli.h line 628:

        bool Exec(const std::vector< std::string >& cmdLine, CliSession& session) override
        {
            if (!IsEnabled()) return false;
            const std::size_t paramSize = sizeof...(Args);
            if (cmdLine.size() != paramSize+1) return false;
            if (Name() == cmdLine[0])
            {
                try
                {
                    auto g = [&](auto ... pars){ func( session.OutStream(), pars... ); };
                    Select<Args...>::Exec(g, std::next(cmdLine.begin()), cmdLine.end());
                }
                catch (std::bad_cast&)
                {
                    return false;
                }
                return true;
            }
            return false;
        }

This is the code that parses a specific command, to see if it's the right one.
As you can see, one of the first line:

     if (cmdLine.size() != paramSize+1) return false;

checks whether the number of arguments matches the number of parameters.
The check should be modified to take into account the optional parameters. This means adding in the `VariadicFunctionCommand` a way to specify whether a parameter is optional and its default value.
Finally, if a parameter is optional, one should change its description to use square brackets instead of angle ones.

Maybe it's feasible. Let me know if you try.
